frontend/expr_inferrer.py

```diff
@@ -0,0 +1,372 @@
+            raise HomogeneousTypesConflict(set_type.get_name(), cur_type.get_name())
+    return TSet(set_type)
+
+def is_numeric(t):
+	return t.is_subtype(TFloat())
+
+def infer_binary_operation(node, context):
+    """Infer the type of binary operations
+
+    Handled cases:
+        - Sequence multiplication, ex: [1,2,3] * 2 --> [1,2,3,1,2,3]
+        - Sequence concatenation, ex: [1,2,3] + [4,5,6] --> [1,2,3,4,5,6]
+        - Arithmatic and bitwise operations, ex: (1 + 2) * 7 & (2 | -123) / 3
+    """
+    left_type = infer(node.left, context)
+    right_type = infer(node.right, context)
+
+    if isinstance(node.op, ast.Mult):
+        # Handle sequence multiplication. Ex.:
+        # [1,2,3] * 2 --> [1,2,3,1,2,3]
+        # 2 * "abc" -- > "abcabc"
+        if left_type.is_subtype(TInt()) and is_sequence(right_type):
+            return right_type
+        elif right_type.is_subtype(TInt()) and is_sequence(left_type):
+            return left_type
+
+    if isinstance(node.op, ast.Add): # Check if it is a concatenation operation between sequences
+        if isinstance(left_type, TTuple) and isinstance(right_type, TTuple):
+            # Handle tuples concatenation:
+            # (1, 2.0, "string") + (True, X()) --> (1, 2.0, "string", True, X())
+            # The result type is the concatenation of both tuples' types
+            new_tuple_types = left_type.types + right_type.types
+            return TTuple(new_tuple_types)
+
+        if is_sequence(left_type) and is_sequence(right_type):
+            if left_type.is_subtype(right_type):
+                return right_type
+            elif right_type.is_subtype(left_type):
+                return left_type
+
+    if isinstance(node.op, ast.Div): # Check if it is a float division operation
+        if left_type.is_subtype(TFloat()) and right_type.is_subtype(TFloat()):
+            return TFloat()
+
+    if is_numeric(left_type) and is_numeric(right_type): # Normal arithmatic or bitwise operation
+        # TODO: Prevent floats from doing bitwise operations
+        if left_type.is_subtype(right_type):
+            return right_type
+        elif right_type.is_subtype(left_type):
+            return left_type
+
+    raise TypeError("Cannot perform operation ({}) on two types: {} and {}".format(type(node.op).__name__, left_type.get_name(), right_type.get_name()))
+
+def infer_unary_operation(node, context):
+    """Infer the type for unary operations
+
+    Examples: -5, not 1, ~2
+    """
+    if isinstance(node.op, ast.Not): # (not expr) always gives bool type
+        return TBool()
+    # TODO: prevent floats from doing the unary operator '~'
+    return infer(node.operand, context)
+
+def infer_if_expression(node, context):
+    """Infer expressions like: {(a) if (test) else (b)}.
+
+    Return a union type of both (a) and (b) types.
+    """
+    a_type = infer(node.body, context)
+    b_type = infer(node.orelse, context)
+
+    if type(a_type) is type(b_type):
+        return a_type
```

Comment:
Could also check for a subtype relationship in either direction (and then return the type that the other is a subtype of). Or have a function create_union_type that takes some types, checks if one of them is a supertype of all others, and returns that type in that case, and otherwise creates a union type. That way, you'd avoid something like Union[str, object], which is identical to object.